### PR TITLE
NAS-114555 / 22.02 / Ensure that zpool handles are freed in iterator

### PR DESCRIPTION
**Pull request by anodos325**

### CODE REVIEW

**Comment by anodos325** [comment]
Copy iterator logic from `property snapshots` and friends. zpool_iter() allocates pool handles that must be closed if we break out of generator.

**Review by yocalebo — APPROVED** [review]
Can confirm, leak without these changes and no leak with them 😄

**Comment by yocalebo** [comment]
Just nits but you have 2 different types of checks. i.e.

```
if self.handle != NULL
```
and
```
if not self.handle
```

I like explicit checks (!= NULL)

**Comment by sonicaj** [comment]
Agreed with @yocalebo :p

**Comment by anodos325** [comment]
Yeah, that bothered me too.

File: libzfs.pyx [hunk 1]
```diff
@@ -1011,19 +1024,39 @@ cdef class ZFS(object):
...
 
             cdef ZFSPool pool
-            pools = []
+            cdef iter_state iter
+            cdef libzfs.zpool_handle_t *handle
 
-            with nogil:
-                libzfs.zpool_iter(self.handle, self.__iterate_pools, <void*>pools)
+            iter.length = 0
+            iter.array = <uintptr_t *>malloc(32 * sizeof(uintptr_t))
```
Comment (ghost) [inline libzfs.pyx]: I believe we should relinquish the GIL while doing this section, too.

Comment (yocalebo) [reply libzfs.pyx]: Good catch. This can actually be refactored quite easily to a single call to `with nogil:`

Comment (yocalebo) [reply libzfs.pyx]: Meh, maybe a little trickier than what I initially thought 😆

**Review by ghost — APPROVED** [review]

**Pull request merged**